### PR TITLE
tlaplus18: init at 1.8.0

### PR DESCRIPTION
**Pull request by mgregson**
This is a pre-release of version 1.8 of TLA+. It adds, among other
things, a TLA+ REPL.

Release Notes: https://github.com/tlaplus/tlaplus/releases/tag/v1.8.0

###### Description of changes

This creates the new package `tlaplus18` providing the pre-release TLA+ 1.8.0. It does so by copying the standard `tlaplus` package from `default.nix` to `tlaplus18.nix` and applying the following modifications:
 - changed `version` to 1.8.0
 - updated `sha256` to match the new SHA
 - added the `tlarepl` wrapper to launch the TLA+ REPL

Note that the 1.7.2 release of TLA+ is 2 years newer than the 1.8.0 release. The 1.8.0 release contains new features that I think are worth providing, however.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by florentc** [comment]
Hello, I have two questions:
- why not updating the existing package from 1.7.2 to 1.8?
- is the alias `tlarepl` for the new `tlc2.REPL` the "official" alias given by other distributions? I am asking because the Nix version of `tlaplus` used to provide aliases that were not the same as what everybody else uses (e.g. `tla2sany` instead of `tlasany`) so I think it would be nice if we made sure everybody uses `tlarepl` (and not another name such as `tlcrepl` for example).

**Comment by mgregson** [comment]
Hello! Thank you for taking the time to look this over!

> * why not update the existing package from 1.7.2 to 1.8?

Version 1.8 of TLA+ is [marked as "pre-release" on the official repository](https://github.com/tlaplus/tlaplus/releases/tag/v1.8.0), so I thought it would be better if I  didn't bring everyone else along on the pre-release adventure with me. There are new features in here (eg, the REPL) that are worth making available though.

If the general consensus is that it's better to upgrade the existing package, I'm happy to do that instead.

> * is the alias `tlarepl` for the new `tlc2.REPL` the "official" alias given by other distributions? I am asking because the Nix version of `tlaplus` used to provide aliases that were not the same as what everybody else uses (e.g. `tla2sany` instead of `tlasany`) so I think it would be nice if we made sure everybody uses `tlarepl` (and not another name such as `tlcrepl` for example).

That's a great question. I'm not sure. When I looked around online for other people packaging tools for TLA+ I found very few, especially packaging the 1.8 release where the REPL was introduced. Even the official website only seems to document releases as recent as 1.4.5. The release notes for version 1.8.0 show how to run the REPL by calling into the class directly using `jar` - you can see that on [the screencast found in the release notes](https://asciinema.org/a/3lsDwbmVG0iyZHJ3RbhNycQS0). The one example I found was [this GitHub repository](https://github.com/pmer/tla-bin) that makes wrappers for accessing TLA+ tools from the command line. It uses `tlarepl` as the command to access the REPL, and the pattern didn't seem to deviate too far from the existing commands in the 1.7.2 package.

All that said, I don't have any investment in how we name the command - I just want to use it. If you think `tlcrepl` would be better I'm happy to make that change.

**Comment by florentc** [comment]
They called it "The Clarke release" and it has a  new version number v1.8.0.
I think we should update the existing package instead. If they release fixes then we will update accordingly to v1.8.x or above as they do. Let's not introduce another package.

As for the name, they seem to call it "an interactive TLA+ REPL" so `tlarepl` seems the best choice indeed (and we could still rename in the future if another name becomes the general usage in the community at some point).

**Review by florentc — CHANGES_REQUESTED** [review]
- Update the existing `tlaplus` package instead

**Comment by mgregson** [comment]
Happy to make the change! Just a note - the 1.7.2 release is actually _newer_ than the 1.8.0 release by 2 full years if you look at the tag dates. I'm not sure what that's about.

I'll update the PR (force push) to replace the 1.7.2 package with the 1.8.0 package, and update the commit message in the commit appropriately.

**Comment by mgregson** [comment]
> * Update the existing `tlaplus` package instead

Done!

**Comment by florentc** [comment]
> Just a note - the 1.7.2 release is actually newer than the 1.8.0 release by 2 full years if you look at the tag dates. I'm not sure what that's about.

Oh yes you are right! This changes everything. In that case you were right to create a new package in the first place. Sorry for the loss of time and energy here.

What they did with version numbering is very confusing. They should have made it clear in the version name itself that this is not an actual v1.8 but an alpha/beta/whatever. I wonder if they dropped it or still are working on it.

**Comment by mgregson** [comment]
Great! I'll revert this PR to its original state and let you know when it's done. I backed up the original variant just in case.

**Comment by mgregson** [comment]
Done!

File: pkgs/applications/science/logic/tlaplus/tlaplus18.nix
Comment (florentc) [inline pkgs/applications/science/logic/tlaplus/tlaplus18.nix:38]: Do you want to add yourself as a maintainer?

**Review by florentc — APPROVED** [review]

File: pkgs/applications/science/logic/tlaplus/tlaplus18.nix
Comment (mgregson) [reply pkgs/applications/science/logic/tlaplus/tlaplus18.nix:38]: I'm not sure it's appropriate for such a small contribution. Perhaps on the next one!

**Comment by mgregson** [comment]
Bringing up to date with the latest master. This has been a pretty long time. Any chance it's going to get merged?

**Review by Janik-Haag — COMMENTED** [review]
LGTM; its up to you if you want to add yourself as maintainer but I would be all for it.

**Comment by Janik-Haag** [comment]
can you squash the commits?

I'm really sorry no committer reviewed this earlier :( 
In the future feel free to ping me for reviews, and please ask questions if you have any.

**Comment by mgregson** [comment]
I will squash the commits and tag you when it's done. By now I've got a [separate MR](https://github.com/NixOS/nixpkgs/pull/231149) on the same package that adds me as a maintainer, so I'll leave that bit there.

**Comment by mgregson** [comment]
@Janik-Haag Squashed and ready to go!

**Review by Janik-Haag — APPROVED** [review]
LGTM; I will merge this as soon as ci is done since some one else approved this before the squashed commits.

**Comment by mgregson** [comment]
It looks like CI failed here for some reason. I didn't make any changes from the previous version, so I'm surprised to see this. I'll make sure to dig in and tag you again when this is resolved. Sorry for the extra iterations!

**Comment by Janik-Haag** [comment]
@ofborg build tlaplus18

**Comment by Janik-Haag** [comment]
Upstream force pfusched something and made a release with the same version number which now obviously has a diffrent hash.. they do be like this sometimes :(

I took the liberty of updating it and pushing the change into the PR no need for you to debug this :D 
Just waiting for CI build and you saying that the update works fine.

**Comment by mgregson** [comment]
Ugh. That's annoying. Thanks for tracking that down!

**Comment by Janik-Haag** [comment]
Did the updated work for you as well and the package still behaves as intended?

**Comment by mgregson** [comment]
@Janik-Haag I tested it out, and everything works! I made one small change: I ported across the changes from #231149 to test it on my darwin-aarch64 machine instead of digging out another computer. Those changes have already been merged to master on the main package, so they shouldn't be controversial here.

I also looked around for a way to fetch the release artifact from a specific commit so we could avoid the release overwriting problem, but didn't find one. Unfortunately, the package will likely break if upstream continues overwriting the artifacts.

**Comment by Janik-Haag** [comment]
> @Janik-Haag I tested it out, and everything works! I made one small change: I ported across the changes from https://github.com/NixOS/nixpkgs/pull/231149 to test it on my darwin-aarch64 machine instead of digging out another computer. Those changes have already been merged to master on the main package, so they shouldn't be controversial here.

Thank's lgtm

> I also looked around for a way to fetch the release artifact from a specific commit so we could avoid the release overwriting problem, but didn't find one. Unfortunately, the package will likely break if upstream continues overwriting the artifacts.

You could use the fetchFromGitHub wrapper which in this case is usually preferred and build it from source, but you don't have to. I personally find it better to use the tag instead of a rev and from my experience upstream doesn't do re-releases to often and it's more the exception.  

> My apologies for the delay in testing this out. It's been a long week.

No worries ^^

**Comment by Janik-Haag** [comment]
@florentc would you mind also giving it a review before I merge it, to make sure I didn't miss anything?

**Comment by Janik-Haag** [comment]
Result of `nixpkgs-review pr 185499` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>tlaplus18</li>
  </ul>
</details>

**Pull request merged**
